base/linalg/dense.jl

```diff
@@ -693,6 +690,309 @@ function inv(A::StridedMatrix{T}) where T
+compute the cosine. Otherwise, the cosine is determined by calling [`exp`](@ref).
+
+# Examples
+```jldoctest
+julia> cos(ones(2, 2))
+2×2 Array{Float64,2}:
+  0.291927  -0.708073
+ -0.708073   0.291927
+```
+"""
+function cos(A::StridedMatrix{<:Real})
+    if issymmetric(A)
+        return full(cos(Symmetric(A)))
+    end
+    return real(exp!(im*A))
+end
+function cos(A::StridedMatrix{<:Complex})
+    if ishermitian(A)
+        return full(cos(Hermitian(A)))
+    end
+    return (exp!(im*A) + exp!(-im*A)) / 2
+end
+
+"""
+    sin(A::StridedMatrix)
+
+Compute the matrix sine of a square matrix `A`.
+
+If `A` is symmetric or Hermitian, its eigendecomposition ([`eigfact`](@ref)) is used to
+compute the sine. Otherwise, the sine is determined by calling [`exp`](@ref).
+
+# Examples
+```jldoctest
+julia> sin(ones(2, 2))
+2×2 Array{Float64,2}:
+ 0.454649  0.454649
+ 0.454649  0.454649
+```
+"""
+function sin(A::StridedMatrix{<:Real})
+    if issymmetric(A)
+        return full(sin(Symmetric(A)))
+    end
+    return imag(exp!(im*A))
+end
+function sin(A::StridedMatrix{<:Complex})
+    if ishermitian(A)
+        return full(sin(Hermitian(A)))
+    end
+    return (exp!(im*A) - exp!(-im*A)) / 2im
+end
+
+"""
+    sincos(A::StridedMatrix)
+
+Compute the matrix sine and cosine of a square matrix `A`.
+
+# Examples
+```jldoctest
+julia> S, C = sincos(ones(2, 2));
+
+julia> S
+2×2 Array{Float64,2}:
+ 0.454649  0.454649
+ 0.454649  0.454649
+
+julia> C
+2×2 Array{Float64,2}:
+  0.291927  -0.708073
+ -0.708073   0.291927
+```
+"""
+function sincos(A::StridedMatrix{<:Real})
+    if issymmetric(A)
+        return full.(sincos(Symmetric(A)))
+    end
+    c, s = reim(exp!(im*A))
+    return s, c
+end
+function sincos(A::StridedMatrix{<:Complex})
+    if ishermitian(A)
+        return full.(sincos(Hermitian(A)))
+    end
+    X, Y = exp!(im*A), exp!(-im*A)
+    return (X - Y) / 2im, (X + Y) / 2
+end
+
+"""
+    tan(A::StridedMatrix)
+
+Compute the matrix tangent of a square matrix `A`.
+
+If `A` is symmetric or Hermitian, its eigendecomposition ([`eigfact`](@ref)) is used to
+compute the tangent. Otherwise, the tangent is determined by calling [`exp`](@ref).
+
+# Examples
+```jldoctest
+julia> tan(ones(2, 2))
+2×2 Array{Float64,2}:
+ -1.09252  -1.09252
+ -1.09252  -1.09252
+```
+"""
+function tan(A::StridedMatrix{<:Real})
+    if issymmetric(A)
+        return full(tan(Symmetric(A)))
+    end
+    s, c = sincos(A)
+    return s / c
+end
+function tan(A::StridedMatrix{<:Complex})
+    if ishermitian(A)
+        return full(tan(Hermitian(A)))
+    end
+    X, Y = exp!(im*A), exp!(-im*A)
+    return (X - Y) / (im * (X + Y))
+end
+
+"""
+    cosh(A::StridedMatrix)
+
+Compute the matrix hyperbolic cosine of a square matrix `A`.
+"""
+function cosh(A::StridedMatrix{T}) where T
+    if ishermitian(A)
+        return full(cosh(Hermitian(A)))
+    end
+    return (exp(A) + exp!(-A)) / 2
```

Comment:
You could avoid allocations here with
```jl
Y = exp(A)
Y .= (Y .+ exp!(-A)) ./ 2
```
But wouldn't it be more efficient to do:
```jl
Y = exp(A)
Y .= (Y .+ inv(Y)) ./ 2
```
since surely `inv` is faster than `exp!`?